stac_validator.py

```diff
@@ -213,76 +214,106 @@ def is_valid_url(self, url):
+                status["catalogs"]["invalid"] = 1
+
+            childs = self._get_childs_urls(stac_content, stac_path)
+
+        elif type(stac_content) is dict and any(
+            field in Collections_Fields for field in stac_content.keys()
         ):
             # Congratulations, It's a Collection!
             # Collections will validate as catalog.
-            self.message["asset_type"] = "collection"
-            self.validate_stac(
-                self.stac_file, cache["catalog-{}".format(self.stac_version)]
+            message["asset_type"] = "collection"
+            is_valid_stac, err_message = self.validate_json(
+                stac_content, cache["catalog-{}".format(self.stac_version)]
             )
 
-            if self.message["valid_stac"]:
-                self.status["collections"]["valid"] += 1
+            message["valid_stac"] = is_valid_stac
+            message["error_message"] = err_message
+
+            if message["valid_stac"]:
+                status["collections"]["valid"] = 1
             else:
-                self.status["collections"]["invalid"] += 1
-            self.message["children"] = await self.validate_catalog_contents()
-        elif "error_type" in self.message:
+                status["collections"]["invalid"] = 1
+
+            childs = self._get_childs_urls(stac_content, stac_path)
+
+        elif "error_type" in message:
             pass
 
         else:
             # Congratulations, It's an Item!
-            self.message["asset_type"] = "item"
-            self.validate_stac(
-                self.stac_file, cache["item-{}".format(self.stac_version)]
+            message["asset_type"] = "item"
+            is_valid_stac, err_message = self.validate_json(
+                stac_content, cache["item-{}".format(self.stac_version)]
             )
+            message["valid_stac"] = is_valid_stac
+            message["error_message"] = err_message
 
-            if self.message["valid_stac"]:
-                self.status["items"]["valid"] += 1
+            if message["valid_stac"]:
+                status["items"]["valid"] = 1
             else:
-                self.status["items"]["invalid"] += 1
+                status["items"]["invalid"] = 1
+
+            childs = []
+
+        message["path"] = stac_path
 
-        self.message["path"] = str(self.fpath)
+        return message, status, childs
+
+    def run(self):
+        """
+        Entry point
+        :return: message json
+
+        """
+        childs = [self.stac_file]
+        while True:
+            concurrent = 10
+            with futures.ThreadPoolExecutor(max_workers=concurrent) as executor:
+                future_tasks = [
+                    executor.submit(self._validate, url) for url in childs
+                ]
+                childs = []
+                for task in futures.as_completed(future_tasks):
+                    message, status, new_childs = task.result()
```

Comment:
each process (`_validate`) returns `message, status, new_childs`. `new_childs` will be processed in the next `while` iteration